### PR TITLE
fix file loading on iOS, expose cache size arguments

### PR DESCRIPTION
**Pull request by lukepighetti**

### CODE REVIEW

**Comment by lukepighetti** [comment]
Friendly ping @999eagle

**Comment by 999eagle** [comment]
Ah sorry, seems like I've missed this PR.. Thanks for the ping :)

File: packages/video_player/lib/video_player.dart [hunk 1]
```diff
@@ -184,8 +194,15 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
...
-  static int maxCacheFileSize = 10 * 1024 * 1024;
+  /// The maximum cache size to keep on disk in bytes.
+  ///
+  /// eg, 100MB is `100 * 1024 * 1024`
+  final int maxCacheSize;
+
+  /// The maximum size of each individual file in bytes.
+  ///
+  /// eg, 10MB is `10 * 1024 * 1024`
+  final int maxCacheFileSize;
```
Comment (999eagle) [inline packages/video_player/lib/video_player.dart]: Why did you change this to be non-static? The backend cache size doesn't change after it has been first created, so I feel like keeping these static creates less problems. Otherwise it might be assumed that the cache sizes can be changed with every newly created `VideoPlayerController`.

Comment (lukepighetti) [reply packages/video_player/lib/video_player.dart]: I wasn't aware that each VideoPlayerController shared the same cache, although it does make sense. Consumers will need to be able to configure this outside of the package source though, so how about we add a comment that states that the cache size is chosen based on the first call? Also, the package had it setup so that the only way to turn off the cache was to set cache size to zero, so that's how I decided it was alright to configure on/off with cache size. We need to be able to disable cache for local File and Asset videos at a minimum, and the static package cache sizes likely wont work for everyone.

Comment (999eagle) [reply packages/video_player/lib/video_player.dart]: Adding a comment would also work. I had intended for consumers to set the static variables at some point during app startup, but you're right, that would make it non-intuitive to turn off caching. Maybe it'd be possible to change the cache sizes dynamically though? I'm not sure if the backends support something like that though.
Files and assets aren't cached anyway as the backends only cache if they're passed a http or https URI.

Comment (lukepighetti) [reply packages/video_player/lib/video_player.dart]: Files wouldn't even play on iOS until I make this PR, fwiw. My solution was to pipe non-cached requests to the old video playing mechanism. I make no judgement on if this is the best way forward, though!

Comment (lukepighetti) [reply packages/video_player/lib/video_player.dart]: Honestly I don't need to do anything but set the cache size once and be able to play File/Asset reliably. This PR achieves that for me personally but I agree it should be more clear to consumers what's going on under the hood.

Comment (999eagle) [reply packages/video_player/lib/video_player.dart]: Ah okay, then I guess I missed that check on iOS. That part of the PR makes sense and looks good :)
The API for cache sizes doesn't really reflect what's going on though. I still think that the static variables should be used for the sizes. Maybe it'd make sense to just use the `bool isCached` argument?

Comment (lukepighetti) [reply packages/video_player/lib/video_player.dart]: But then you lose the ability to configure the cache size. I think the problem here is iOS (maybe Android?) use singleton pattern for the cache but the package does not. So I think it's okay to just add a comment that the first cache size called is the one that gets used for the entire app.

Comment (999eagle) [reply packages/video_player/lib/video_player.dart]: Why do you lose that ability? A consumer can just set these variables once in during their app's startup instead of making sure to always pass non-zero values to `VideoPlayerController`.

Both backends use a singleton cache, maybe it would also be useful to add a new platform method call for configuring this instead of always passing the cache sizes on each `create` call, but that would be another thing.

Comment (lukepighetti) [reply packages/video_player/lib/video_player.dart]: Are you imagining maybe a static method on VideoPlayerController like .setCacheSize? I can see that working. But you'd only be able to call it once as of today and I think that's fine. But when you said use static variables to set cache I thought you meant static constants within the package, not sure what you meant by static variables.

Personally I'm more of the "use it if it works" camp and today it appears the PR in front of you has all the functionality we're discussing its just that people need to know that the first time a cache size is set it won't change.

For my uses, I would never need to change the cache size, only set it, and I don't mind if that is implicit which it is in this PR.

Comment (999eagle) [reply packages/video_player/lib/video_player.dart]: Hm a static method would work as well and would be cleaner to be honest. Static variables were what I was using before this PR, they're not constants. These two variables:
```dart
  static int maxCacheSize = 100 * 1024 * 1024;
  static int maxCacheFileSize = 10 * 1024 * 1024;
```
They could be set by the consumer using `VideoPlayerController.maxCacheSize = 2000 * 1024 * 1024` for example if someone needs a particularly large cache.
What I meant was a new call over the platform channel to the platform-specific implementation, though. That way we also change the communication between Dart and platform code to be more clear. Still, this is something for a future change.

Yeah, that's true. It does all it needs to do. I'm still not sure that passing cache sizes each time to the `VideoPlayerController` constructor is better than the static variables I previously used.

Comment (lukepighetti) [reply packages/video_player/lib/video_player.dart]: Honestly this is a pattern I've never seen before and didn't even know this was possible. It looks like you're correct that it would work as you intend, but it is an informal pattern that may not be expected by Dart consumers, as far as I know. Generally speaking I think people expect static methods/properties to be pure without side effects, but maybe that's just me? 🤔 

```dart
main(){
  final foo = Foo();
  print(Foo.number); /// 1
  print(foo.currentNumber); /// 1
  
  Foo.number = 2;
  print(Foo.number); /// 2
  print(foo.currentNumber); /// 2
}

class Foo {
  static int number = 1;
  int get currentNumber => number;
}
```

 I've only see this used to create singleton classes with an `instance` getter. But I can't come up with a reason you shouldn't use it as long as it's well documented. 

My reason for configuring it in the constructor was mostly to be able to turn off the cache. I only have one place in my app where I use the cache and so specifying it there has worked well for me but I understand it might not work well for everyone. Turning off the cache was necessary to fix an issue where iOS wouldn't load videos from a File.

Comment (999eagle) [reply packages/video_player/lib/video_player.dart]: I mean, setting the static variable doesn't have any side effects in itself. It only changes what's being read later on. I don't know the Dart ecosystem enough to know which patterns are usually used and which aren't though, so if you say that this is an unexpected way to do things, then maybe static variables shouldn't be used. I still would like some way for a consumer to set the cache sizes just once and not pass them to every single constructor call. Maybe with a static method instead to make it at least a bit more clear how it's intended to be used?
Turning off the cache could still be realised using the `enableCache` flag you've added in this PR instead of always giving the option to pass new cache sizes. The current change also has the problem that a consumer could call the `VideoPlayerController` constructor with cache sizes and `enableCache = false` because they don't want to cache this one Uri.

Comment (lukepighetti) [reply packages/video_player/lib/video_player.dart]: My vote would be for a "setCache" static method and back it up internally with private static variables and have it throw an error if it's called twice. Thoughts?

Comment (lukepighetti) [reply packages/video_player/lib/video_player.dart]: I just posted one up that I haven't fooled around with yet.

Comment (999eagle) [reply packages/video_player/lib/video_player.dart]: That looks pretty good, I like the idea of actually throwing an error there!
I guess now all that remains is to remove the two cache size parameters in the `VideoPlayerController.network` constructor and then we're good.

Comment (lukepighetti) [reply packages/video_player/lib/video_player.dart]: Woops! Nice catch. Fixed that and added some asserts on setCacheSize

Comment (999eagle) [reply packages/video_player/lib/video_player.dart]: Okay I think it's good now! Thanks :)

Comment (lukepighetti) [reply packages/video_player/lib/video_player.dart]: Yay!

**Comment by lukepighetti** [comment]
@999eagle can you open up issues on this repo? I am having more problems with URLs parsing on iOS. This url does not work with `feature/caching` but does work with `video_player`. I don't have much experience with objc so any assistance is appreciated. It appears that my attempt at encoding URLs is not working with firebasestorage, it appears to have very specific encoding requirements. I'm not quite sure what to do.

https://firebasestorage.googleapis.com/v0/b/skatex-8d802.appspot.com/o/timeline-videos%2Finstagram%3A15239448540%2F055b574c.mp4?alt=media&token=c6da1e33-f3ff-41e2-a6f0-bdb475a2f6d9

some reading that I will dig into tomorrow: https://stackoverflow.com/questions/32242712/replacement-for-stringbyaddingpercentescapesusingencoding-in-ios9/44643893
